icu4c/source/i18n/unitconverter.cpp

```diff
@@ -15,6 +18,126 @@
 U_NAMESPACE_BEGIN
 
 namespace {
+
+/* Internal Structure */
+
+enum Constants {
+    CONSTANT_FT2M,    // ft2m stands for foot to meter.
+    CONSTANT_PI,      // PI
+    CONSTANT_GRAVITY, // Gravity
+    CONSTANT_G,
+    CONSTANT_GAL_IMP2M3, // Gallon imp to m3
+    CONSTANT_LB2KG,      // Pound to Kilogram
+
+    // Must be the last element.
+    CONSTANTS_COUNT
+};
+
+typedef enum SigNum {
+    NEGATIVE = -1,
+    POSITIVE = 1,
+} SigNum;
+
+/* Represents a conversion factor */
+struct Factor {
+    double factorNum = 1;
+    double factorDen = 1;
+    double offset = 0;
+    bool reciprocal = false;
+    int32_t constants[CONSTANTS_COUNT] = {};
+
+    void multiplyBy(const Factor &rhs) {
+        factorNum *= rhs.factorNum;
+        factorDen *= rhs.factorDen;
+        for (int i = 0; i < CONSTANTS_COUNT; i++)
+            constants[i] += rhs.constants[i];
+
+        offset += rhs.offset;
+    }
+
+    void divideBy(const Factor &rhs) {
+        factorNum *= rhs.factorDen;
+        factorDen *= rhs.factorNum;
+        for (int i = 0; i < CONSTANTS_COUNT; i++)
+            constants[i] -= rhs.constants[i];
+
+        offset += rhs.offset;
+    }
+
+    // Apply the power to the factor.
+    void power(int32_t power) {
+        // multiply all the constant by the power.
+        for (int i = 0; i < CONSTANTS_COUNT; i++)
+            constants[i] *= power;
+
+        bool shouldFlip = power < 0; // This means that after applying the absolute power, we should flip
+                                     // the Numerator and Denomerator.
+
+        factorNum = std::pow(factorNum, std::abs(power));
+        factorDen = std::pow(factorDen, std::abs(power));
+
+        if (shouldFlip) {
+            // Flip Numerator and Denomirator.
+            std::swap(factorNum, factorDen);
+        }
+    }
+
+    // Flip the `Factor`, for example, factor= 2/3, flippedFactor = 3/2
+    void flip() {
+        std::swap(factorNum, factorDen);
+
+        for (int i = 0; i < CONSTANTS_COUNT; i++) {
+            constants[i] *= -1;
+        }
+    }
+
+    // Apply SI prefix to the `Factor`
+    void applySiPrefix(UMeasureSIPrefix siPrefix) {
+        if (siPrefix == UMeasureSIPrefix::UMEASURE_SI_PREFIX_ONE) return; // No need to do anything
+
+        double siApplied = std::pow(10.0, std::abs(siPrefix));
+
+        if (siPrefix < 0) {
+            factorDen *= siApplied;
+            return;
+        }
+
+        factorNum *= siApplied;
+    }
+};
+
+/* Helpers */
+
+using icu::double_conversion::StringToDoubleConverter;
+
+// Returns `double` from a scientific number(i.e. "1", "2.01" or "3.09E+4")
+double strToDouble(StringPiece strNum) {
+    // We are processing well-formed input, so we don't need any special options to
+    // StringToDoubleConverter.
+    StringToDoubleConverter converter(0, 0, 0, "", "");
+    int32_t count;
+    return converter.StringToDouble(strNum.data(), strNum.length(), &count);
```

Comment:
If `count < strNum.length()`, it means that the number failed to parse fully.  I suggest adding a `UErrorCode&` parameter and setting it in this condition, perhaps to `U_INVALID_FORMAT_ERROR`.